File: i2f-jdk/i2f-swl/web/i2f-swl/src/i2f-turbo-web/i2f-core/codec/Base64Obfuscator.js
/**
 * Base64混淆工具
 */
import Random from '../util/Random'

const Base64Obfuscator = {
    /**
     * @const
     * @return {String}
     */
    OBF_PREFIX() {
        return '$.'
    },
    /**
     *
     * @param bs4 {String}
     * @param prefix {boolean}
     * @return String
     */
    encode(bs4, prefix) {
        let builder = ''
        if (prefix) {
            builder += this.OBF_PREFIX()
        }
        let jpc = (bs4.length + Random.nextLowerInt(31)) % 10
        builder += (String.fromCharCode('0'.charCodeAt(0) + jpc))
        for (let p = 0; p < jpc; p++) {
            const ird = Random.nextLowerInt(16)
            if (ird < 10) {
                builder += (String.fromCharCode('0'.charCodeAt(0) + ird))
            } else {
                builder += (String.fromCharCode('A'.charCodeAt(0) + (ird - 10)))
            }
        }
        jpc = jpc % 2 + 2
        for (let i = 0; i < bs4.length; i++) {
            if (i % jpc == 0 && bs4.charAt(i) != '=') {
                const ird = Random.nextLowerInt(16)
                if (ird < 10) {
                    builder += (String.fromCharCode('0'.charCodeAt(0) + ird))
                } else {
                    builder += (String.fromCharCode('A'.charCodeAt(0) + (ird - 10)))
                }
            }
            builder += (bs4.charAt(i))
        }
        return builder
    },
    /**
     *
     * @param sob {String}
     * @return {String}
     */
    decode(sob) {
        let str = sob
        if (sob.startsWith(this.OBF_PREFIX())) {
            str = sob.substring(this.OBF_PREFIX().length)
        }
        let builder = ''
        let jpc = str.charAt(0).charCodeAt(0) - '0'.charCodeAt(0)
        str = str.substring(1 + jpc)
        jpc = jpc % 2 + 3
        let i = 0
        while (i < str.length) {
            if (i % jpc == 0 && str.charAt(i) != '=') {
                i++
                continue
            }
            builder += (str.charAt(i))
            i++
        }
        return builder
    }
}

export default Base64Obfuscator

File: i2f-jdk/i2f-codec-impl/src/main/java/i2f/codec/str/obfuscate/Base64Obfuscator.java
package i2f.codec.str.obfuscate;


import java.security.SecureRandom;

/**
 * @author Ice2Faith
 * @date 2022/5/6 15:27
 * @desc base64 混淆器实现
 * 每隔原始的base64串2位，随机插入一个hex的char
 */
public class Base64Obfuscator {
    public static final String OBF_PREFIX = "$.";

    public static String encode(String bs4, boolean prefix) {
        if (bs4 == null || bs4.isEmpty()) {
            return bs4;
        }
        SecureRandom random = new SecureRandom();
        StringBuilder builder = new StringBuilder();
        if (prefix) {
            builder.append(OBF_PREFIX);
        }
        int jpc = (bs4.length() + random.nextInt(31)) % 10;
        builder.append((char) ('0' + jpc));
        for (int p = 0; p < jpc; p++) {
            int ird = random.nextInt(16);
            if (ird < 10) {
                builder.append((char) ('0' + ird));
            } else {
                builder.append((char) ('A' + (ird - 10)));
            }
        }
        jpc = jpc % 2 + 2;
        for (int i = 0; i < bs4.length(); i++) {
            if (i % jpc == 0 && bs4.charAt(i) != '=') {
                int ird = random.nextInt(16);
                if (ird < 10) {
                    builder.append((char) ('0' + ird));
                } else {
                    builder.append((char) ('A' + (ird - 10)));
                }
            }
            builder.append(bs4.charAt(i));
        }
        return builder.toString();
    }

    public static String decode(String sob) {
        if (sob == null || sob.isEmpty()) {
            return sob;
        }
        String str = sob;
        if (sob.startsWith(OBF_PREFIX)) {
            str = sob.substring(OBF_PREFIX.length());
        }
        StringBuilder builder = new StringBuilder();
        int jpc = str.charAt(0) - '0';
        str = str.substring(1 + jpc);
        jpc = jpc % 2 + 3;
        int i = 0;
        while (i < str.length()) {
            if (i % jpc == 0 && str.charAt(i) != '=') {
                i++;
                continue;
            }
            builder.append(str.charAt(i));
            i++;
        }
        return builder.toString();
    }

}
